### ISSUE TITLE
Possible bug with methods-to-keep-client-default-value

### ISSUE THREAD

**Issue by vvladmirs** [Azure/autorest.csharp#3780]
When `swagger.json` contains `x-ms-client-default`, the methods listed in `methods-to-keep-client-default-value` are ignored and the method parameters are not set to default values when client generated.

`swagger.json`:

```json
"parameters": [
    {
    "name": "X-HTTP-Method-Override",
    "in": "header",
    "required": true,
    "schema": {
        "type": "string",
        "default": "DELETE",
        "x-ms-client-default": "DELETE"
        }
    }
],
```

<details>
<summary>Full `swagger.json`</summary>

```json
{
  "openapi": "3.0.1",
  "info": {
    "title": "n",
    "version": "v1"
  },
  "paths": {
    "/api/structure/attributes ": {
      "post": {
        "tags": [
          "Attributes"
        ],
        "summary": "Delete attributes",
        "description": "Following permissions are required: **Attribute_Delete**",
        "operationId": "Attributes_DeleteAttributes",
        "parameters": [
          {
            "name": "X-HTTP-Method-Override",
            "in": "header",
            "required": true,
            "schema": {
              "type": "string",
              "default": "DELETE",
              "x-ms-client-default": "DELETE"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "array",
                "items": {
                  "type": "string",
                  "format": "uuid"
                }
              }
            },
            "text/json": {
              "schema": {
                "type": "array",
                "items": {
                  "type": "string",
                  "format": "uuid"
                }
              }
            },
            "application/*+json": {
              "schema": {
                "type": "array",
                "items": {
                  "type": "string",
                  "format": "uuid"
                }
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Success"
          },
          "400": {
            "description": "Bad Request",
            "content": {
              "text/plain": {
                "schema": {
                  "$ref": "#/components/schemas/ProblemDetails"
                }
              },
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ProblemDetails"
                }
              },
              "text/json": {
                "schema": {
                  "$ref": "#/components/schemas/ProblemDetails"
                }
              }
            }
          }
        },
        "security": [
          {
            "OAuth2": [
              "ns"
            ]
          }
        ]
      }
    }
  },
  "components": {
    "schemas": {
      "ProblemDetails": {
        "type": "object",
        "properties": {
          "type": {
            "type": "string",
            "nullable": true
          },
          "title": {
            "type": "string",
            "nullable": true
          },
          "status": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          },
          "detail": {
            "type": "string",
            "nullable": true
          },
          "instance": {
            "type": "string",
            "nullable": true
          }
        },
        "additionalProperties": { }
      }
    },
    "securitySchemes": {
      "OAuth2": {
        "type": "oauth2",
        "flows": {
          "implicit": {
            "authorizationUrl": "/connect/authorize",
            "scopes": {
              "ns": "Common scope for all Web-API"
            }
          },
          "password": {
            "tokenUrl": "/connect/token",
            "scopes": {
              "ns": "Common scope for all Web-API"
            }
          },
          "authorizationCode": {
            "authorizationUrl": "/connect/authorize",
            "tokenUrl": "/connect/token",
            "scopes": {
              "ns": "Common scope for all Web-API"
            }
          }
        }
      }
    }
  }
}
```

</details>

Command line options for client generation:

```bash
dotnet build -c Release /t:GenerateCode
```

Client configuration:

``` yaml
input-file: .\swagger.json
public-clients: true
title: NC
generation1-convenience-client: true
skip-csproj: true
clear-output-folder: true
eol: lf

use-extension:
  "@autorest/csharp": "latest"

security: AADToken
security-header-name: OAuth2
security-scopes: 
  - "ns"
  - "offline_access"

methods-to-keep-client-default-value:
  - "Attributes_DeleteAttributes"

```

## Expected Behavior

Generated method in the client:

```c#
public virtual Response<ProblemDetails> DeleteAttributes(string xHttpMethodOverride = "DELETE", ...)
```

## Current Behavior

```c#
public virtual Response<ProblemDetails> DeleteAttributes(string xHttpMethodOverride, ...)
```

<details>
<summary>Full generated method</summary>

```c#
        /// <summary> Delete attributes. </summary>
        /// <param name="xHttpMethodOverride"> The String to use. The default value is "DELETE". </param>
        /// <param name="body"> The ArrayOfPostContentSchemaItem to use. </param>
        /// <param name="cancellationToken"> The cancellation token to use. </param>
        /// <remarks> Following permissions are required: **Attribute_Delete**. </remarks>
        public virtual Response<ProblemDetails> DeleteAttributes(string xHttpMethodOverride, IEnumerable<Guid> body = null, CancellationToken cancellationToken = default)
        {
            using var scope = _clientDiagnostics.CreateScope("AttributesClient.DeleteAttributes");
            scope.Start();
            try
            {
                return RestClient.DeleteAttributes(xHttpMethodOverride, body, cancellationToken);
            }
            catch (Exception e)
            {
                scope.Failed(e);
                throw;
            }
        }
```

</details>

## Environment

```
MSBuild version 17.7.1+971bf70db for .NET
AutoRest code generation utility [cli version: 3.6.2; node: v16.19.0]
AutoRest core 3.9.7
AutoRest/csharp 3.0.0-beta.20230924.2
AutoRest/modelerfour 4.26.0
```

This issue is reproduced from commit [77106fdbc328ae7920e7e8a0089df901489d9352](https://github.com/Azure/autorest.csharp/commit/77106fdbc328ae7920e7e8a0089df901489d9352)

**Comment by chunyu3** [comment]
Hello @vvladmirs .NET SDK will not honor `x-ms-client-default-value`, and will not set this default value in method signature. `methods-to-keep-client-default-value` is an exception to avoid breaking change for existing GAed SDKs. And it is not give out a way to switch honor `x-ms-client-default-value` or not. 
And since HLC does not has breaking changes when eliminate the default value, so we does not implement this exception for HLC.

@vvladmirs , Do you have any special service which need default value?

**Comment by vvladmirs** [comment]
Sorry for the late reply.
Yes, we have the following case:
we have a header, with a single possible value, which in the OpenAPI schema is specified in the `default` parameter (and in `x-ms-client-default-value`).
In the current version, how can I set the default value for a parameter in the client?

Previously, we specified the value of the parameter in `x-ms-client-default-value` and it was set as the default value, for example see **Expected Behavior** above.

In the current version, how can I set the default value for a parameter in the client?
